### PR TITLE
Ssh watchdog client

### PR DESCRIPTION
**Pull request by antonrohr**
This should be pretty straight forward, the first commit is some small refactoring, the second holds the `StartWatchdog` function

### CODE REVIEW

File: OrbitQt/servicedeploymanager.cpp [hunk 1]
```diff
@@ -335,6 +336,25 @@ outcome::result<void> ServiceDeployManager::ConnectToServer() {
   return outcome::success();
 }
 
+void ServiceDeployManager::StartWatchdog(std::string_view start_passphrase,
+                                         int interval_in_ms) {
+  orbit_service_task_->Write(start_passphrase);
+
+  const auto watchdog_send_timer = QPointer(new QTimer());
```
Comment (beckerhe) [inline OrbitQt/servicedeploymanager.cpp:343]: I think it makes more sense to put the timer into a member variable. No allocation necessary.

Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:343]: If not the timeout callback might be called when the service deploy manager is already destructed.

File: OrbitQt/servicedeploymanager.cpp [hunk 2]
```diff
@@ -335,6 +336,25 @@ outcome::result<void> ServiceDeployManager::ConnectToServer() {
...
+
+  QObject::connect(watchdog_send_timer, &QTimer::timeout,
```
Comment (beckerhe) [inline OrbitQt/servicedeploymanager.cpp:345]: Can you use the 4 argument overload of QObject::connect?

File: OrbitQt/servicedeploymanager.h [hunk 3]
```diff
@@ -69,6 +69,8 @@ class ServiceDeployManager : public QObject {
       OrbitSshQt::SftpOperation::FileMode dest_mode);
   outcome::result<void> StopSftpChannel(OrbitSshQt::SftpChannel*);
 
+  void StartWatchdog(std::string_view start_passphrase = "start_watchdog",
```
Comment (beckerhe) [inline OrbitQt/servicedeploymanager.h:72]: Nit: Could you put the text into a kWtachdogStartPassphrase variable into the implementation file?

File: OrbitQt/servicedeploymanager.h [hunk 4]
```diff
@@ -69,6 +69,8 @@ class ServiceDeployManager : public QObject {
...
+                     int interval_in_ms = 1000);
```
Comment (beckerhe) [inline OrbitQt/servicedeploymanager.h:73]: Nit: Also this value?

File: OrbitQt/servicedeploymanager.cpp (hunk 1 above)
Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:343]: By the way the QPointer makes no sense here because watchdog_send_timer only exists for a short time.

**Comment by beckerhe** [comment]
You really don't like references, do you? :sweat_smile:

File: OrbitQt/servicedeploymanager.cpp (hunk 2 above)
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.cpp:345]: Yes, I was wondering what the difference is?

Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:345]: Ah sorry. You're right. This overload is fine. Just leave it like so.

File: OrbitQt/servicedeploymanager.cpp (hunk 1 above)
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.cpp]: makes sense! done

File: OrbitQt/servicedeploymanager.h (hunk 3 above)
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.h]: done

File: OrbitQt/servicedeploymanager.h (hunk 4 above)
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.h]: done

File: OrbitQt/servicedeploymanager.cpp [hunk 5]
```diff
@@ -335,6 +336,25 @@ outcome::result<void> ServiceDeployManager::ConnectToServer() {
...
+                   [this, watchdog_send_timer]() {
+                     if (orbit_service_task_) {
+                       orbit_service_task_->Write(".");
+                     } else {
+                       watchdog_send_timer->stop();
+                       watchdog_send_timer->deleteLater();
+                     }
+                   });
+
+  watchdog_send_timer->start(interval_in_ms);
```
[the remark this answers is not in the document]
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.cpp]: done

**Comment by beckerhe** [comment]
> I actually do like references a lot, but my clang-tidy with more google style things does not like them at all

The restrictions on references in the Google style guide have recently been leveraged. Maybe the clang-tidy rules do not reflect that yet.

File: OrbitQt/servicedeploymanager.cpp [hunk 6]
```diff
@@ -335,6 +337,23 @@ outcome::result<void> ServiceDeployManager::ConnectToServer() {
   return outcome::success();
 }
 
+void ServiceDeployManager::StartWatchdog() {
+  orbit_service_task_->Write(kSshWatchdogPassphrase);
+
+  // ssh_watchdog_timer_ = QPointer(new QTimer());
```
Comment (beckerhe) [inline OrbitQt/servicedeploymanager.cpp:343]: I think you missed removing that line.

File: OrbitQt/servicedeploymanager.cpp [hunk 7]
```diff
@@ -335,6 +337,23 @@ outcome::result<void> ServiceDeployManager::ConnectToServer() {
...
+  QObject::connect(&ssh_watchdog_timer_, &QTimer::timeout, [this]() {
+    if (orbit_service_task_) {
+      orbit_service_task_->Write(".");
+    } else {
+      ssh_watchdog_timer_.stop();
+    }
+  });
+
+  ssh_watchdog_timer_.setInterval(kSshWatchdogInterval);
+  ssh_watchdog_timer_.start();
```
Comment (beckerhe) [inline OrbitQt/servicedeploymanager.cpp:354]: Nit: If you're removing the comment anyway, you could also make one line out of this. ssh_watchdog_timer_.start(kSshWatchdogInterval). But not important at all. If your prefer the two line version it's fine as well.

**Comment by beckerhe** [comment]
> > I actually do like references a lot, but my clang-tidy with more google style things does not like them at all
> 
> The restrictions on references in the Google style guide have recently been leveraged. Maybe the clang-tidy rules do not reflect that yet.

Not leveraged. Lifted I mean!

File: OrbitQt/servicedeploymanager.cpp (hunk 6 above)
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.cpp:343]: true, thanks, that was sloppy

File: OrbitQt/servicedeploymanager.cpp (hunk 7 above)
Comment (antonrohr) [reply OrbitQt/servicedeploymanager.cpp]: I actually missed that overload in the documentation. Its split up in signals and normal member functions. Anyway, sloppy again on my side

**Comment by antonrohr** [comment]
> The restrictions on references in the Google style guide have recently been lifted

Good to know!

**Pull request merged**